server/controllers/auth.js

```diff
@@ -0,0 +1,61 @@
+const User = require("../models/User");
+const ToDoList = require("../models/ToDoList");
+const ToDo = require("../models/ToDo");
+const {StatusCodes} = require("http-status-codes");
+const {BadRequestError, UnauthenticatedError} = require("../errors");
+
+const register = async (req, res) => {
+  const user = await User.create({...req.body});
+  const token = user.createJWT();
+  res
+    .status(StatusCodes.CREATED)
+    .json({user: {username: user.username}, token});
+};
+
+const login = async (req, res) => {
+  const {username, password} = req.body;
+
+  if (!username || !password) {
+    throw new BadRequestError("Please provide username and password.");
+  }
+  const user = await User.findOne({username});
+  if (!user) {
```

Comment:
Please elaborate.

Reply:
My original comment was really vague, apologies. I was referring specifically to line 23 (which is part of this if(!user) function) "throw new UnauthenticatedError("Invalid....")". On line 27, you throw another UnauthenticatedError for an incorrect password. I think there is a way we could combine these two functions:
if (!user) ... [line 22]
and 
if (!isPasswordCorrect) ... [line 26]
into one function  - or perhaps use the checkUsernameAndPassword function you referred to in your response on line 18.

Reply:
A `checkUsernameAndPassword` function is implemented in `/server/controllers/auth.js` in the coming `server-sprint-2` branch that is called inside both the `login` and `deactivate` controller functions  (replacing original lines 16-28 and 35-47 with the function calls) for purpose of validation. 

However, there're more than just the number of lines and characters to consider in writing DRY code, which is trumped by performance considerations: factors such as time and memory complexities, call stack and queue, etc. In short, if a user has entered invalid username, you want to return the error without wasting the resource to check whether the password is valid. You proposal would require waiting after checking the validity of both the username and password, when the user entered an invalid username, to return the error--chipping that bit of performance for writing 3 fewer lines of code. 

 Also, in case approval on review is required for merging, as here, I would not withhold it without proposing a precise solution for the suspected deficiencies to which I raise attention.